Introduction to Free Monads/article.md

```diff
@@ -0,0 +1,702 @@
+
+### Trees as free monads
+
+To close this section out, let's look at another application of free monads.
+Namely, let's build a binary search tree from a sorted list.
+
+Our base functor for the binary tree would look like this:
+
+```haskell
+data BinTreeF l a = NodeF l a a
+  deriving Functor
+```
+
+You might ask yourself, where are the leaf and/or nil branch constructors.
+We can actually encode those as the "return value" of type `Maybe l`, so we don't really need them here.
+
+Now we define our free monad type and build the tree.
+
+```haskell
+type FreeBinTree l = Free (BinTreeF l)
+
+buildBalanced :: [Int] -> FreeBinTree Int (Maybe Int)
+buildBalanced [] = pure Nothing
+buildBalanced [x] = pure $ Just x
+buildBalanced xs = do
+  let len = length xs
+      (l,x:r) = splitAt (len `div` 2) xs
+  b <- liftF $ NodeF x l r
+  buildBalanced b
+```
+
+The first two cases, where lists have sizes of zero and one respectively, are self-evident.
+If there are more elements in the list, we divide it in half.
+The approximate "middle" element serves as the current node value.
+We then recursively construct left and right subtrees from the corresponding halves of the list.
+
+Notice how we're passing left and right halves of the list to `NodeF` and then bind it to name `b`.
+As we discussed previously, multiple functor variables per constructor encode non-determinism.
+So here we basically split the computation into two branches.
+We then recursively build the tree for each branch.
+
+The caveat here is there's no simple way to print this all, and working with a tree wrapped in the `Free` is probably not particularly convenient.
+So we can convert this `FreeBinTree` into a regular binary tree, for example:
+
+```haskell
+data BinTree l = Nil | Leaf l | Branch l (BinTree l) (BinTree l)
+  deriving Show
+
+convert :: FreeBinTree a (Maybe a) -> BinTree a
+convert (Pure Nothing) = Nil
+convert (Pure (Just x)) = Leaf x
+convert (Free f) =
+  let BranchF x l r = convert <$> f
+  in Branch x l r
+```
+
+Note how `Pure` values correspond to leaves, and `Free` values correspond to branches.
+Running `convert . buildBalanced` on a list produces the expected result, i.e. a binary search tree.
+
+```
+> convert $ buildBalanced [0..10]
+5 (2 (1 0
+        Nil)
+     (4 3
+        Nil))
+  (8 (7 6
+        Nil)
+     (10 9
+         Nil))
+```
+
+## Conclusions
+
+> The only way to learn a new programming language is by writing programs in it.
+>
+> -- Dennis Ritchie
+
+Hopefully, after this brief introduction, you understand, at least in principle, what free monads are, and how they may be useful.
+Of course, the only way to get proficient with them is to use them in projects, and no blog post can replace actual hands-on experience, but that applies to everything in programming.
```

Comment:
Are there some exercises that you could add? We are big on exercises at Serokell Blog. :D

Reply:
:thinking: I could come up with some variations on the examples described here, but those would be somewhat boring. I could also trim the examples section somewhat, and present what I trim off as exercises, however that would diminish the value of the article on its own (i.e. for someone who doesn't care much for exercises). I'll see if I can come up with something more interesting, though.

FWIW, I do have one exercise in the middle of the theoretic section (i.e. equivalence proof of monad laws defined via bind vs Kleisli composition), but honestly I don't expect anyone to actually do it  :sweat_smile:

Reply:
Boring is fine, usually exercises are at the end and they are very skippable, but whomever will need it -- implement a free monad instance for this or smth like that will be just in time. :)